frame/merkle-mountain-range/rpc/src/lib.rs

```diff
@@ -109,7 +112,7 @@ pub trait MmrApi<BlockHash> {
 	fn generate_proof(
 		&self,
 		leaf_index: LeafIndex,
-		at: Option<BlockHash>,
+		best_known_block: Option<BlockNumber>,
 	) -> RpcResult<LeafProof<BlockHash>>;
```

Comment:
Let's not change runtime `MmrApi` and simply add new `mmr_generateHistoricalProof` API (avoiding a few lines of code duplication is certainly not worth dealing with changing the runtime API). Same for RPC API.

Also, I would keep `at: Option<BlockHash>,` in all runtime api functions for consistency, even if it's redundant for the `mmr_generateHistoricalProof` API. Calling runtime APIs `at` different block states is part of the runtime API "contract", I don't think `sp_api::impl_runtime_apis!` will even work without that param.

Reply:
Ok. Thanks ! I'll add `mmr_generateHistoricalProof` then.

Reply:
> Let's not change runtime `MmrApi` and simply add new `mmr_generateHistoricalProof` API (avoiding a few lines of code duplication is certainly not worth dealing with changing the runtime API version - I'm also not in favor of overloading existing API to also handle "historic" case, makes the API harder to use IMO). Same for RPC API.

I fully agree with @acatangiu on this given the tradeoff cost applicable at the time.
Incorporating yesterday's decision to break this API for the leaf->block transition in #12339 though, I want to revisit this affected design tradeoff while the API will have to change anyway: preserving API version seemed the primary motivation for the code duplication, and that's no longer weighing against this.

Now the tradeoff reduces to
a. keep equivalent code in multiple locations in sync, vs
b. pass an extra `None` to `mmr_generateProof` when proving for the `best_block`

so the balance is different.

A unified API via `b.` does indeed make the API more verbose when dealing with the `best_block` only, but given that a light client's `best_known_block` will typically lag the reference chain's `best_block`, I expect it's less prone to errors by API consumers if they need to be deliberate here when actually wanting to track the `best_block`.

Reply:
Personally I still prefer b. And since we're changing the API anyway I was also wondering if we should keep `generate_proof()`, `verify_proof()`, and `verify_proof_stateless()` since all these can be done via batching methods (e.g. `generate_batch_proof()`, etc). I think it would be best to discuss all these in a separate issue.

Reply:
I created #12391 for continuing this discussion

Reply:
I agree, let’s discuss it separately